lib/liquid/translators/tags/case.ex

```diff
@@ -0,0 +1,156 @@
+  end
+
+  def translate([value, {:clauses, clauses}, {:else, else_tag_values}]) do
+    create_block_for_case(Markup.literal(value), clauses, else_tag_values)
+  end
+
+  def translate([value, {:else, else_tag_values}]) do
+    create_block_for_case_else(Markup.literal(value), else_tag_values)
+  end
+
+  def translate([value, badbody, {:clauses, clauses}]) do
+    create_block_for_case(Markup.literal(value), badbody, clauses)
+  end
+
+  def translate([value, badbody, {:clauses, clauses}, {:else, else_tag_values}]) do
+    create_block_for_case(Markup.literal(value), badbody, clauses, else_tag_values)
+  end
+
+  def translate([value, badbody, {:else, else_tag_values}]) do
+    create_block_for_case_else(Markup.literal(value), badbody, else_tag_values)
+  end
+
+  def translate([value, badbody]) do
+    block = %Liquid.Block{name: :case, markup: Markup.literal(value), nodelist: [badbody]}
+    to_case_block(block)
+  end
+
+  defp when_to_nodelist({:when, [conditions: [head | tail], body: values]})
+       when is_bitstring(head) do
+    tag = %Liquid.Tag{
+      name: :when,
+      markup: "\"#{Markup.literal(head)}\"" <> Markup.literal(tail)
+    }
+
+    result = Liquid.NimbleTranslator.process_node(values)
+    [tag, result]
+  end
+
+  defp when_to_nodelist({:when, [conditions: conditions, body: values]}) do
+    tag = %Liquid.Tag{
+      name: :when,
+      markup: Enum.join(conditions)
+    }
+
+    result = Liquid.NimbleTranslator.process_node(values)
+    [tag, result]
+  end
+
+  defp else_tag(values) do
+    process_list = Liquid.NimbleTranslator.process_node(values)
+
+    if is_list(process_list) do
+      else_liquid_tag = %Liquid.Tag{
+        name: :else
+      }
+
+      [else_liquid_tag | process_list]
+    else
+      else_liquid_tag = %Liquid.Tag{
+        name: :else
+      }
+
+      [else_liquid_tag, process_list]
+    end
+  end
+
+  defp create_block_for_case(markup, when_alone) do
+    nodelist = Enum.map(when_alone, &when_to_nodelist/1) |> List.flatten()
+    block = %Liquid.Block{name: :case, markup: markup, nodelist: nodelist}
+    to_case_block(block)
+  end
+
+  defp create_block_for_case(markup, when_tag, else_tag_values) when is_list(when_tag) do
+    nodelist = Enum.map(when_tag, &when_to_nodelist/1) |> List.flatten()
+    nodelist_plus_else = [nodelist | else_tag(else_tag_values)] |> List.flatten()
+    block = %Liquid.Block{name: :case, markup: markup, nodelist: nodelist_plus_else}
+    to_case_block(block)
+  end
+
+  defp create_block_for_case(markup, badbody, when_tag) do
+    nodelist_when = Enum.map(when_tag, &when_to_nodelist/1) |> List.flatten()
+    full_list = [badbody | nodelist_when] |> List.flatten()
+    block = %Liquid.Block{name: :case, markup: markup, nodelist: full_list}
+    to_case_block(block)
+  end
+
+  defp create_block_for_case(markup, badbody, when_tag, else_tag_values) do
+    nodelist_when = Enum.map(when_tag, &when_to_nodelist/1) |> List.flatten()
```

Comment:
Is this flatten necessary? We flatten again on line 135.

Reply:
corrected in the previous comment